cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/schedule/ProgramSchedule.java

```diff
@@ -52,7 +52,7 @@ public ProgramSchedule(String name, String description,
                          ProgramId programId, Map<String, String> properties,
                          Trigger trigger, List<? extends Constraint> constraints, long timeoutMillis) {
     this.description = description;
-    this.programId = programId;
+    this.programId = programId.toDefaultVersion();
```

Comment:
Why is this needed? Isn't it easier to just ignore version throughout the scheduler code?

Reply:
For `ProgramSchedule` this is not needed necessarily, during operations like listSchedule(programId), we will just need to compare all the other details expect versionId.  I was thinking making all programId "versionless" so we could do the comparison more easily.
However, this doesn't work for `ProgramStatusTrigger` which is another table - `triggerStore`. Because it has this string triggerkey concept that has programId in it (for example `programStatus:program:default.FakeApp.-SNAPSHOT.workflow.DataPipelineWorkflow.completed`), this trigger key is used to scan the table to find out which if there is any pipeline needs to be triggered on completion of `FakeApp.DataPipelineWorkflow`.  Here we cannot ignore the versionId without impacting query performance as it is part of the string,  but I am happy to know if there is a better workaround.

Reply:
the trigger key is a detail internal to the store (which class is the triggerStore?). We can always insert a constant `-SNAPSHOT` as the version just in that store, instead of modifying all the callers to pass in a ProgramId with `-SNAPSHOT` version.

That being said, it's much cleaner to replace all the ProgramIds in the trigger/schedule classes with ProgramReference, is that happening in a separate PR?

Reply:
good idea to just insert that constant in store, will try it out.
Yes replacing programId with ProgramReference will come in another PR

Reply:
triggerStore belongs to ProgramScheduleStore

Reply:
@albertshau I feel this is pretty much the minimum changes that can be made for schedule/trigger to work. Always inserting "-SNAPSHOT" resolves the triggering key prefix issue, but in a few places where it is comparing the `ProgramId` object directly, we had to make the version consistent.  Do you think we can go with this and replace `programId` with `ProgramReference` in a following PR (after https://github.com/cdapio/cdap/pull/14671 is merged)?